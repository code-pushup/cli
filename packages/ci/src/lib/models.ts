import type { SourceFileIssue } from './issues';
import type { MonorepoTool } from './monorepo';

/**
 * Customization options for {@link runInCI}
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#options
 */
export type Options = {
  monorepo?: boolean | MonorepoTool;
  projects?: string[] | null;
  task?: string;
  bin?: string;
  config?: string | null;
  directory?: string;
  silent?: boolean;
  debug?: boolean;
  detectNewIssues?: boolean;
  logger?: Logger;
};

/**
 * {@link Options} with filled-in defaults.
 */
export type Settings = Required<Options>;

/**
 * Branches/commits for {@link runInCI}
 */
export type GitRefs = {
  head: GitBranch;
  base?: GitBranch;
};

/**
 * API client for {@link runInCI} - adapter for CI provider
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#provider-api-client
 */
export type ProviderAPIClient = {
  maxCommentChars: number;
  downloadReportArtifact?: () => Promise<string | null>;
  listComments: () => Promise<Comment[]>;
  updateComment: (id: number, body: string) => Promise<Comment>;
  createComment: (body: string) => Promise<Comment>;
};

/**
 * PR comment from {@link ProviderAPIClient}
 */
export type Comment = {
  id: number;
  body: string;
  url: string;
};

/**
 * Git ref (e.g. 'main') and commit sha (e.g. '9d7568265bd5b17d1bb51a6eb6407e0d387058d2')
 */
export type GitBranch = {
  ref: string;
  sha: string;
};

/**
 * Logger instance (e.g. `console`) for reporting progress and problems
 */
export type Logger = {
  error: (message: string) => void;
  warn: (message: string) => void;
  info: (message: string) => void;
  debug: (message: string) => void;
};

/**
 * Resolved return value of {@link runInCI}
 */
export type RunResult = StandaloneRunResult | MonorepoRunResult;

/**
 * Resolved return value of {@link runInCI} in standalone mode
 */
export type StandaloneRunResult = Omit<ProjectRunResult, 'name'> & {
  mode: 'standalone';
  commentId?: number;
};

/**
 * Resolved return value of {@link runInCI} in monorepo mode
 */
export type MonorepoRunResult = {
  mode: 'monorepo';
  projects: ProjectRunResult[];
  commentId?: number;
  diffArtifact?: ArtifactData;
};

/**
 * Result of {@link runInCI} for a single project
 */
export type ProjectRunResult = {
  name: string;
  artifacts: {
    report: ArtifactData;
    diff?: ArtifactData;
  };
  newIssues?: SourceFileIssue[];
};

/**
 * Paths to output files from {@link runInCI}, for uploading as job artifact
 */
export type ArtifactData = {
  rootDir: string;
  files: string[];
};
