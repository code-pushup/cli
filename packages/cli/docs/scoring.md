# Scoring

**Scoring strategies** are mathematical approaches used to convert raw metrics into normalized scores
(typically 0-1 range) that provide meaningful feedback about performance against targets or budgets. These
strategies are particularly valuable in software development contexts where you need to track and visualize how
well various metrics (like bundle sizes, test coverage, API latency, or build times) are performing relative to
established thresholds or baselines. Each strategy offers different characteristics - some provide linear
relationships between the metric and score, while others offer more nuanced behaviors like soft caps or
logarithmic scaling.

## Value Scoring Strategies

### Overview Table

| **Strategy Name**     | **Formula**                                                                                                                            | **Description**                                                                |
| --------------------- | -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------ |
| **Percent Used**      | $$\text{score}= \max\!\left(0,\,1 - \frac{S}{M}\right)$$                                                                               | Linearly falls from 1→0 as **S** goes 0→**M**, then clamps at 0.               |
| **Linear Overshoot**  | $$\text{score}= \begin{cases}1,&S\le M\\[4pt]1 - \dfrac{S - M}{M},&S>M\end{cases}$$                                                    | Holds 1 until **S**=**M**, then falls linearly to 0 by **S**=2·**M**.          |
| **Relative Baseline** | $$\text{score}= \mathrm{clamp}_{0..1}\!\Bigl(0.5 + \frac{\text{baseline} - S}{2\,\text{baseline}}\Bigr)$$                              | Centers at 0.5 when **S** matches baseline; >0.5 improvement, <0.5 regression. |
| **Sigmoid Soft-Cap**  | $$\text{score}= \frac{1}{1 + e^{\,k\,(S - M)}}$$                                                                                       | Remains near 1 before **M**, then gently decays, steep after threshold.        |
| **Logarithmic Decay** | $$\text{score}= \max\!\Bigl(0,\,1 - \frac{\log_{10}(1+S)}{\log_{10}(1 + M\,k)}\Bigr)$$                                                 | Penalises large **S** sharply, small overruns have mild impact.                |
| **Tiered Grading**    | $$\text{score}= \begin{cases}1,&P\ge0.9\\[2pt]0.75,&P\ge0.75\\[2pt]0.5,&P\ge0.5\\[2pt]0,&\text{otherwise}\end{cases},\;P=\frac{S}{M}$$ | Maps **S/M** into A/B/C/F bands for clear grade visualization.                 |

---

### Percent Used Details

Directly reflects remaining budget fraction.

**Pros:**

- Intuitive (% headroom).
- Simple to compute.  
  **Cons:**
- Flatlines at 0 once budget is exceeded.

**Implementation:**

```ts
function percentUsed(current: number, max: number): number {
  return Math.max(0, 1 - current / max);
}
```

**When to use it:**  
Ideal for any strict size or time budget (bundle size, build time) where you care about how much headroom remains. Simple dashboards and alerts benefit from a direct “% remaining” metric.

**Examples:**

- **Bundle Stats:** Track JS bundle headroom under a 500 KB limit.
- **Test Coverage:** % lines covered vs 80 % threshold.
- **Page Load (LCP):** Remaining ms under a 2 s budget.

**Chart:**

```mermaid
xychart-beta
    title "Percent Used (M=10)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Percent [1,0.9,0.8,0.7,0.6,0.5,0.4,0.3,0.2,0.1,0]
```

---

### Linear Overshoot Details

Penalty grows linearly with overshoot.

**Pros:**

- Proportional and predictable.
- Easy to explain to stakeholders.  
  **Cons:**
- Can go negative if not clamped.

**Implementation:**

```ts
function linearOvershoot(current: number, max: number): number {
  if (current <= max) {
    return 1;
  }
  return Math.max(0, 1 - (current - max) / max);
}
```

**When to use it:**  
When you want to maintain proportional visibility into how far a metric has exceeded its limit (e.g., memory usage, API latency). Works well if you clamp at 0 to avoid negatives.

**Examples:**

- **API Latency:** Overshoot above 200 ms threshold.
- **Memory Usage:** RAM above configured max.
- **CSS Sizes:** Style sheets above a 50 KB cap.

**Chart:**

```mermaid
xychart-beta
    title "Linear Overshoot"
    x-axis [0,5,10,12,14,16,18,20]
    y-axis "Score" 0 --> 1
    line Overshoot [1,1,1,0.8,0.6,0.4,0.2,0]
```

---

### Relative Baseline Details

Compares current value to a historical baseline.

**Pros:**

- Symmetric: 0.5 = “no change.”
- Highlights improvements/regressions.  
  **Cons:**
- Requires reliable baseline data.

**Implementation:**

```ts
function relativeBaseline(current: number, baseline: number): number {
  const score = 0.5 + (baseline - current) / (2 * baseline);
  return Math.max(0, Math.min(1, score));
}
```

**When to use it:**  
Perfect for tracking trends across releases (bundle size, test performance, code complexity). Shifts focus from absolute budgets to continuous improvement.

**Examples:**

- **Bundle Stats:** Compare today’s bundle size vs last release.
- **Test Duration:** Measure speedup/regression of test suite.
- **Doc Coverage:** Track improvements in storybook coverage.

**Chart:**

```mermaid
xychart-beta
    title "Relative Baseline (baseline=10)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Delta [1,0.95,0.9,0.85,0.8,0.75,0.7,0.65,0.6,0.55,0.5]
```

---

### Sigmoid Soft-Cap Details

Soft threshold with tunable steepness.

**Pros:**

- Gentle near the cap; harsh after.
- Never negative or >1.  
  **Cons:**
- Requires selecting a constant **k**.
- Less intuitive without plotting.

**Implementation:**

```ts
function sigmoidSoftCap(current: number, max: number, k: number = 0.5): number {
  return 1 / (1 + Math.exp(k * (current - max)));
}
```

**Parameter Details:**

- **k** (steepness factor): Controls how quickly the score drops after exceeding the threshold
  - `k = 0.1`: Very gentle decline, scores remain high even with significant overshoot
  - `k = 0.5`: Moderate decline (default), balanced between tolerance and penalty
  - `k = 1.0`: Steep decline, harsh penalty for any overshoot
  - `k = 2.0`: Very steep decline, almost binary behavior near the threshold

**When to use it:**  
Use when minor breaches are acceptable but you want strong pressure against large overruns (e.g. accessibility violations, deprecation warnings).

**Examples:**

- **ESLint Errors:** Small numbers of lint errors tolerated.
- **TS Diagnostics:** Minor type mismatches vs critical errors.
- **Security Audits:** Few low-severe alerts OK; many not.

**Chart:**

```mermaid
xychart-beta
    title "Sigmoid Soft-Cap (M=5, k=0.5)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Sigmoid [0.92,0.88,0.82,0.73,0.62,0.50,0.38,0.27,0.18,0.12,0.08]
```

---

### Logarithmic Decay Details

Log-scale penalty for wide-range data.

**Pros:**

- Highlights massive overruns.
- Small ones barely noticed.  
  **Cons:**
- Formula less intuitive.

**Implementation:**

```ts
function logarithmicDecay(current: number, max: number, k: number = 2): number {
  const score = 1 - Math.log10(1 + current) / Math.log10(1 + max * k);
  return Math.max(0, score);
}
```

**Parameter Details:**

- **k** (scaling factor): Controls the logarithmic compression range
  - `k = 1`: Minimal compression, behaves more like linear decay
  - `k = 2`: Moderate compression (default), good balance for most metrics
  - `k = 5`: Strong compression, very tolerant of small overruns but penalizes large ones
  - `k = 10`: Maximum compression, almost flat until very large overruns

**When to use it:**  
Ideal for metrics spanning orders of magnitude (e.g. network RTT, bundle gzipped vs uncompressed sizes).

**Examples:**

- **CWV (FID):** First Input Delay distribution.
- **Bundle Stats:** Large vs small chunks on log scale.
- **API Throttling:** High‐tail latency events.

**Chart:**

```mermaid
xychart-beta
    title "Logarithmic Decay (M=5, k=2)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Log [1,0.71,0.54,0.42,0.33,0.25,0.19,0.13,0.08,0.04,0]
```

---

### Tiered Grading Details

Discrete grade bands (A/B/C/F).

**Pros:**

- Very readable.
- No continuous math needed for viewers.  
  **Cons:**
- Abrupt jumps hide nuance.

**Implementation:**

```ts
function tieredGrading(current: number, max: number): number {
  const ratio = current / max;
  if (ratio >= 0.9) return 1;
  if (ratio >= 0.75) return 0.75;
  if (ratio >= 0.5) return 0.5;
  return 0;
}
```

**When to use it:**  
Best for executive dashboards or compliance checks where letter‐grade is preferred over percentages.

**Examples:**

- **Test Coverage:** A: ≥90 %, B: 75 – 89 %, C: 50 – 74 %.
- **Doc Coverage:** Storybook or JSDoc rates.
- **Security Score:** CVSS banding.

**Chart:**

```mermaid
xychart-beta
    title "Tiered Grading"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Grade [0,0,0,0,0.5,0.5,0.5,0.75,1,1,1]
```
