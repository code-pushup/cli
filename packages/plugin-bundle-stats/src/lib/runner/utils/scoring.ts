import type { BundleStatsConfig } from '../types.js';

/**
 * Calculates the bundle score based on total size and threshold
 *
 * @param total - Total bundle size in bytes
 * @param config - Bundle configuration containing thresholds
 * @returns Score as a ratio (0-1) of actual size to threshold, capped at 1
 */
export function calculateScore(
  total: number,
  config: BundleStatsConfig,
): number {
  const threshold = config.thresholds.totalSize;
  let maxThreshold: number;

  if (Array.isArray(threshold)) {
    maxThreshold = threshold[1]; // Use max value from [min, max] tuple
  } else {
    maxThreshold = threshold;
  }

  // Score calculation: 1.0 if within threshold, decreasing as size increases
  if (total <= maxThreshold) {
    return 1.0;
  }

  // Calculate how much over the threshold we are
  const excess = total - maxThreshold;
  const excessRatio = excess / maxThreshold;

  // Score decreases linearly, but doesn't go below 0
  return Math.max(0, 1 - excessRatio);
}
