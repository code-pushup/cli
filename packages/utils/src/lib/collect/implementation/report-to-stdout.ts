import { Report } from '@quality-metrics/models';
import chalk from 'chalk';
import cliui from 'cliui';
import {
  countWeightedRefs,
  reportHeadlineText,
  reportOverviewTableHeaders,
  sumRefs,
} from './utils';
import { NEW_LINE } from './md';

const ui = cliui({ width: 60 }); // @TODO check display width

export function reportToStdout(report: Report): void {
  reportToHeaderSection(report);
  reportToMetaSection(report);
  console.log(NEW_LINE);
  reportToOverviewSection(report);
  console.log(NEW_LINE);
  reportToDetailSection(report);
  console.log(NEW_LINE);
  console.log('Made with ❤️ by code-pushup.dev');
}

function reportToHeaderSection(report: Report): void {
  const { packageName, version } = report;
  console.log(`${chalk.bold(reportHeadlineText)} - ${packageName}@${version}`);
}

function reportToMetaSection(report: Report): void {
  const { date, duration, version, packageName, plugins } = report;
  const _print = (text: string) => console.log(chalk.italic(chalk.gray(text)));

  _print(`---`);
  _print(`Package Name: ${packageName}`);
  _print(`Version: ${version}`);
  _print(
    `Commit: feat(cli): add logic for markdown report - 7eba125ad5643c2f90cb21389fc3442d786f43f9`,
  );
  _print(`Date: ${new Date(date).toString()}`);
  _print(`Duration: ${duration}ms`);
  _print(`Plugins: ${plugins?.length}`);
  _print(
    `Audits: ${plugins?.reduce((sum, { audits }) => sum + audits.length, 0)}`,
  );
  _print(`---`);
}

function reportToOverviewSection(report: Report): void {
  const base = {
    width: 20,
    padding: [0, 1, 0, 1],
  };

  // table header
  ui.div(...reportOverviewTableHeaders.map(text => ({ text, ...base })));

  // table content
  report.categories.forEach(({ title, refs }) => {
    const score = sumRefs(refs).toString();
    const audits = `${refs.length.toString()}/${countWeightedRefs(refs)}`;

    ui.div(
      {
        text: `${title}`,
        ...base,
      },
      {
        text: score,
        ...base,
      },
      {
        text: audits,
        ...base,
      },
    );
  });

  console.log(ui.toString());
}

function reportToDetailSection(report: Report): void {
  const { categories, plugins } = report;

  categories.forEach(category => {
    const { title, refs } = category;

    console.log(chalk.bold(`${title} ${sumRefs(refs)}`));

    refs.forEach(
      ({ slug: auditSlugInCategoryRefs, weight, plugin: pluginSlug }) => {
        const audit = plugins
          .find(({ slug }) => slug === pluginSlug)
          ?.audits.find(
            ({ slug: auditSlugInPluginAudits }) =>
              auditSlugInPluginAudits === auditSlugInCategoryRefs,
          );

        if (audit) {
          let content = `${audit.description}` + NEW_LINE;
          if (audit.docsUrl) {
            content += `  ${audit.docsUrl} ${NEW_LINE}`;
          }
          console.log(`- ${audit.title} (${weight})`);
          console.log(`  ${content}`);
        } else {
          // this should never happen
          throw new Error(`No audit found for ${auditSlugInCategoryRefs}`);
        }
      },
    );
  });
}
