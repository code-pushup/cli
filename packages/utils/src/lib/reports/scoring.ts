import {
  AuditReport,
  CategoryConfig,
  CategoryRef,
  Group,
  GroupRef,
  PluginReport,
  Report,
} from '@code-pushup/models';
import { deepClone } from '../transform';

export type EnrichedAuditReport = AuditReport & { plugin: string };
export type WeighedAuditReport = EnrichedAuditReport & { weight: number };
export type EnrichedScoredGroupWithAudits = EnrichedScoredGroup & {
  audits: AuditReport[];
};
export type ScoredCategoryConfig = CategoryConfig & { score: number };

export type EnrichedScoredGroup = Group & {
  plugin: string;
  score: number;
};

export type ScoredReport = Omit<Report, 'plugins' | 'categories'> & {
  plugins: (Omit<PluginReport, 'audits' | 'groups'> & {
    audits: EnrichedAuditReport[];
    groups: EnrichedScoredGroup[];
  })[];
  categories: ScoredCategoryConfig[];
};

export function calculateScore<T extends { weight: number }>(
  refs: T[],
  scoreFn: (ref: T) => number,
): number {
  const { numerator, denominator } = refs.reduce(
    (acc, ref) => {
      const score = scoreFn(ref);
      return {
        numerator: acc.numerator + score * ref.weight,
        denominator: acc.denominator + ref.weight,
      };
    },
    { numerator: 0, denominator: 0 },
  );
  // No division by 0, otherwise we produce NaN
  // This can be caused by:
  // - empty category refs
  // - categories with refs only containing weight of `0`
  // both should get caught when validating the model
  if (!numerator && !denominator) {
    throw new Error(
      '0 division for score. This can be caused by refs only weighted with 0 or empty refs',
    );
  }
  return numerator / denominator;
}

export function scoreReport(report: Report): ScoredReport {
  const allScoredAuditsAndGroups = new Map();

  const scoredPlugins = report.plugins.map(plugin => {
    const { slug, audits, groups } = plugin;

    const updatedAudits = audits.map(audit => ({ ...audit, plugin: slug }));

    updatedAudits.forEach(audit => {
      allScoredAuditsAndGroups.set(`${slug}-${audit.slug}-audit`, audit);
    });

    function groupScoreFn(ref: GroupRef) {
      const score = allScoredAuditsAndGroups.get(
        `${slug}-${ref.slug}-audit`,
      )?.score;
      if (score == null) {
        throw new Error(
          `Group has invalid ref - audit with slug ${slug}-${ref.slug}-audit not found`,
        );
      }
      return score;
    }

    const scoredGroups =
      groups?.map(group => ({
        ...group,
        score: calculateScore(group.refs, groupScoreFn),
        plugin: slug,
      })) ?? [];

    scoredGroups.forEach(group => {
      allScoredAuditsAndGroups.set(`${slug}-${group.slug}-group`, group);
    });

    return { ...plugin, audits: updatedAudits, groups: scoredGroups };
  });

  function catScoreFn(ref: CategoryRef) {
    const key = `${ref.plugin}-${ref.slug}-${ref.type}`;
    const item = allScoredAuditsAndGroups.get(key);
    if (!item) {
      throw new Error(
        `Category has invalid ref - ${ref.type} with slug ${key} not found in ${ref.plugin} plugin`,
      );
    }
    return item.score;
  }

  const scoredCategories = report.categories.map(category => ({
    ...category,
    score: calculateScore(category.refs, catScoreFn),
  }));

  return {
    ...deepClone(report),
    plugins: scoredPlugins,
    categories: scoredCategories,
  };
}
