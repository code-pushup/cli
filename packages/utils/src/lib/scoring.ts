import {
  AuditGroup,
  AuditGroupRef,
  AuditReport,
  CategoryConfig,
  CategoryRef,
  PluginReport,
  Report,
} from '@code-pushup/models';
import { deepClone } from './transformation';

export type EnrichedAuditReport = AuditReport & { plugin: string };
type ScoredCategoryConfig = CategoryConfig & { score: number };

export type EnrichedScoredAuditGroup = AuditGroup & {
  plugin: string;
  score: number;
};

export type ScoredReport = Omit<Report, 'plugins' | 'categories'> & {
  plugins: (Omit<PluginReport, 'audits' | 'groups'> & {
    audits: EnrichedAuditReport[];
    groups: EnrichedScoredAuditGroup[];
  })[];
  categories: ScoredCategoryConfig[];
};

export function calculateScore<T extends { weight: number }>(
  refs: T[],
  scoreFn: (ref: T) => number,
): number {
  const { numerator, denominator } = refs.reduce(
    (acc, ref) => {
      const score = scoreFn(ref);
      return {
        numerator: acc.numerator + score * ref.weight,
        denominator: acc.denominator + ref.weight,
      };
    },
    { numerator: 0, denominator: 0 },
  );
  return numerator / denominator;
}

export function scoreReport(report: Report): ScoredReport {
  const scoredReport = deepClone(report) as ScoredReport;
  const allScoredAuditsAndGroups = new Map();

  scoredReport.plugins?.forEach(plugin => {
    const { audits } = plugin;
    const groups = plugin.groups || [];

    audits.forEach(audit => {
      const key = `${plugin.slug}-${audit.slug}-audit`;
      audit.plugin = plugin.slug;
      allScoredAuditsAndGroups.set(key, audit);
    });

    function groupScoreFn(ref: AuditGroupRef) {
      const score = allScoredAuditsAndGroups.get(
        `${plugin.slug}-${ref.slug}-audit`,
      )?.score;
      if (score == null) {
        throw new Error(
          `Group has invalid ref - audit with slug ${plugin.slug}-${ref.slug}-audit not found`,
        );
      }
      return score;
    }

    groups.forEach(group => {
      const key = `${plugin.slug}-${group.slug}-group`;
      group.score = calculateScore(group.refs, groupScoreFn);
      group.plugin = plugin.slug;
      allScoredAuditsAndGroups.set(key, group);
    });
    plugin.groups = groups;
  });

  function catScoreFn(ref: CategoryRef) {
    const key = `${ref.plugin}-${ref.slug}-${ref.type}`;
    const item = allScoredAuditsAndGroups.get(key);
    if (!item) {
      throw new Error(
        `Category has invalid ref - ${ref.type} with slug ${key} not found in ${ref.plugin} plugin`,
      );
    }
    return item.score;
  }

  const scoredCategoriesMap = new Map();
  // eslint-disable-next-line functional/no-loop-statements
  for (const category of scoredReport.categories) {
    category.score = calculateScore(category.refs, catScoreFn);
    scoredCategoriesMap.set(category.slug, category);
  }

  scoredReport.categories = Array.from(scoredCategoriesMap.values());

  return scoredReport;
}
