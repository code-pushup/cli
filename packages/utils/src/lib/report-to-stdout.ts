import cliui from '@isaacs/cliui';
import chalk from 'chalk';
import Table from 'cli-table3';
import { NEW_LINE } from './md';
import {
  CODE_PUSHUP_DOMAIN,
  FOOTER_PREFIX,
  countCategoryAudits,
  formatReportScore,
  reportHeadlineText,
  reportRawOverviewTableHeaders,
  sortAudits,
} from './report';
import { ScoredReport } from './scoring';

function addLine(line = ''): string {
  return line + NEW_LINE;
}

export function reportToStdout(report: ScoredReport): string {
  let output = '';

  output += addLine(reportToHeaderSection(report));
  output += addLine();
  output += addLine(reportToDetailSection(report));
  output += addLine(reportToOverviewSection(report));
  output += addLine(`${FOOTER_PREFIX} ${CODE_PUSHUP_DOMAIN}`);

  return output;
}

function reportToHeaderSection(report: ScoredReport): string {
  const { packageName, version } = report;
  return `${chalk.bold(reportHeadlineText)} - ${packageName}@${version}`;
}

function reportToOverviewSection({
  categories,
  plugins,
}: ScoredReport): string {
  let output = addLine(chalk.magentaBright.bold('Categories'));
  output += addLine();

  const table = new Table({
    head: reportRawOverviewTableHeaders,
    colAligns: ['left', 'right', 'right'],
    style: {
      head: ['cyan'],
    },
  });

  table.push(
    ...categories.map(({ title, score, refs }) => [
      title,
      withColor({ score }),
      countCategoryAudits(refs, plugins),
    ]),
  );

  output += addLine(table.toString());

  return output;
}

function reportToDetailSection(report: ScoredReport): string {
  const { plugins } = report;

  let output = '';

  plugins.forEach(({ title, audits }) => {
    output += addLine();
    output += addLine(chalk.magentaBright.bold(`${title} audits`));
    output += addLine();

    const ui = cliui({ width: 80 });

    audits.sort(sortAudits).forEach(({ score, title, displayValue, value }) => {
      ui.div(
        {
          text: withColor({ score, text: '‚óè' }),
          width: 2,
          padding: [0, 1, 0, 0],
        },
        {
          text: title,
          padding: [0, 3, 0, 0],
        },
        {
          text: chalk.cyanBright(displayValue || `${value}`),
          width: 10,
          padding: [0, 0, 0, 0],
        },
      );
    });

    output += addLine(ui.toString());
    output += addLine();
  });

  return output;
}

function withColor({ score, text }: { score: number; text?: string }) {
  let str = text ?? formatReportScore(score);
  const style = text ? chalk : chalk.bold;
  if (score < 0.5) {
    str = style.red(str);
  } else if (score < 0.9) {
    str = style.yellow(str);
  } else {
    str = style.green(str);
  }
  return str;
}
