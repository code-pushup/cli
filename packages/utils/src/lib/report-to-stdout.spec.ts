import { describe } from 'vitest';
import { report } from '@code-pushup/models/testing';
import { reportToStdout } from './report-to-stdout';
import { scoreReport } from './scoring';

describe('report-to-stdout', () => {
  it('should contain all sections when using the fixture report', () => {
    const logOutput = reportToStdout(scoreReport(report()));
    // logOutput.replace(/\u001B\[\d+m/g, '') removes all color codes from the output
    // for snapshot readability
    // eslint-disable-next-line no-control-regex
    expect(logOutput.replace(/\u001B\[\d+m/g, '')).toMatchSnapshot();
  });
});
