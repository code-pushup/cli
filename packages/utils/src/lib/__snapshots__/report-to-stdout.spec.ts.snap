// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`report-to-stdout > should contain all sections when using the fixture report 1`] = `
"Code PushUp Report - @code-pushup/core@0.0.1


ESLint audits

● Disallow assignment operators in conditional expressions            0
● Disallow reassigning \`const\` variables                              0
● Disallow the use of \`debugger\`                                      0
● Disallow invalid regular expression strings in \`RegExp\` construct   0
  ors
● Disallow the use of undeclared variables unless mentioned in \`/*g   0
  lobal */\` comments
● Disallow loops with a body that allows only one iteration           0
● Disallow negating the left operand of relational operators          0
● Disallow use of optional chaining in contexts where the \`undefine   0
  d\` value is not allowed
● Disallow unused variables                                           1 warning
● Require calls to \`isNaN()\` when checking for \`NaN\`                  0
● Enforce comparing \`typeof\` expressions against valid strings        0
● Require braces around arrow function bodies                         1 warning
● Enforce camelcase naming convention                                 0
● Enforce consistent brace style for all control statements           0
● Require the use of \`===\` and \`!==\`                                  1 warning
● Enforce a maximum number of lines of code in a function             1 warning
● Enforce a maximum number of lines per file                          0
● Disallow variable declarations from shadowing variables declared    3 warnings
  in the outer scope
● Require \`let\` or \`const\` instead of \`var\`                           0
● Require or disallow method and property shorthand syntax for obje   3 warnings
  ct literals
● Require using arrow functions for callbacks                         0
● Require \`const\` declarations for variables that are never reassig   1 warning
  ned after declared
● Disallow using Object.assign with an object literal as the first    0
  argument and prefer the use of object spread instead
● Require or disallow \\"Yoda\\" conditions                               0
● Disallow missing \`key\` props in iterators/collection literals       1 warning
● Disallow missing props validation in a React component definition   6 warnings
● Disallow missing React when using JSX                               0
● enforces the Rules of Hooks                                         0
● verifies the list of dependencies for Hooks like useEffect and si   2 warnings
  milar
● Disallow missing displayName in a React component definition        0
● Disallow comments from being inserted as text nodes                 0
● Disallow duplicate properties in JSX                                0
● Disallow \`target=\\"_blank\\"\` attribute without \`rel=\\"noreferrer\\"\`     0
● Disallow undeclared variables in JSX                                0
● Disallow React to be incorrectly marked as unused                   0
● Disallow variables used in JSX to be incorrectly marked as unused   0
● Disallow passing of children as props                               0
● Disallow when a DOM element is using both children and dangerousl   0
  ySetInnerHTML
● Disallow usage of deprecated methods                                0
● Disallow direct mutation of this.state                              0
● Disallow usage of findDOMNode                                       0
● Disallow usage of isMounted                                         0
● Disallow usage of the return value of ReactDOM.render               0
● Disallow using string references                                    0
● Disallow unescaped HTML entities from appearing in markup           0
● Disallow usage of unknown DOM property                              0
● Enforce ES5 or ES6 class for returning value in render function     0


Lighthouse audits

● First Contentful Paint                                              1.2 s
● Largest Contentful Paint                                            1.5 s
● Total Blocking Time                                                 0 ms
● Cumulative Layout Shift                                             0
● Speed Index                                                         1.2 s


Categories

┌────────────────┬───────┬────────┐
│ Category       │ Score │ Audits │
├────────────────┼───────┼────────┤
│ Performance    │    92 │      6 │
├────────────────┼───────┼────────┤
│ Bug prevention │    68 │     16 │
├────────────────┼───────┼────────┤
│ Code style     │    54 │     13 │
└────────────────┴───────┴────────┘

Made with ❤ by code-pushup.dev
"
`;
